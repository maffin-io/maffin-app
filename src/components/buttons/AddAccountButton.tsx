import React from 'react';
import Modal from '@/components/Modal';
import { BiPlusCircle } from 'react-icons/bi';

import { useDataSource } from '@/hooks';
import AccountForm from '@/components/forms/account/AccountForm';

export default function AddAccountButton(): JSX.Element {
  const { save } = useDataSource();
  const [isModalOpen, setIsModalOpen] = React.useState<boolean>(false);

  return (
    <>
      <Modal
        title="Add account"
        open={isModalOpen}
        setOpen={setIsModalOpen}
      >
        <AccountForm
          onSave={() => {
            save();
            setIsModalOpen(false);
          }}
        />
      </Modal>
      <button
        type="button"
        className="btn-primary"
        onClick={() => setIsModalOpen(!isModalOpen)}
      >
        <BiPlusCircle className="mr-1" />
        Add Account
      </button>
    </>
  );
}
