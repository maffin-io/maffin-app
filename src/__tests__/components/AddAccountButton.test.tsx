import React from 'react';
import {
  waitFor,
  render,
  screen,
  fireEvent,
} from '@testing-library/react';

import AddAccountButton from '@/components/AddAccountButton';
import type { AccountFormProps } from '@/components/forms/account/AccountForm';
import type { ModalProps } from '@/components/Modal';

jest.mock('@/components/Modal', () => {
  function Modal(props: ModalProps) {
    return (
      <div data-testid="modal" className="Modal">
        <span>{props.setOpen.name}</span>
        <span>{`open: ${props.open}`}</span>
        <span>{props.title}</span>
        {props.children}
      </div>
    );
  }

  return Modal;
});

jest.mock('@/components/forms/account/AccountForm', () => {
  function AccountForm(props: AccountFormProps) {
    return (
      <div className="AccountForm">
        <span>{props.onSave.name}</span>
      </div>
    );
  }

  return AccountForm;
});

describe('AddAccountButton', () => {
  it('opens modal when clicking the button', async () => {
    render(
      <AddAccountButton
        onSave={jest.fn()}
      />,
    );

    const button = await screen.findByRole('button', { name: /add account/i });
    fireEvent.focus(button);

    fireEvent.click(button);
    await waitFor(() => {
      screen.getByTestId('modal');
    });

    expect(screen.getByTestId('modal')).toMatchSnapshot();
  });
});
