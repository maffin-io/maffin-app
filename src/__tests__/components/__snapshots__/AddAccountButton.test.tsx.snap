// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddAccountButton opens modal when clicking the button 1`] = `
<div
  class="Modal"
  data-testid="modal"
>
  <span>
    bound dispatchSetState
  </span>
  <span>
    open: true
  </span>
  <span>
    Add account
  </span>
  <div
    class="AccountForm"
  >
    <span>
      onSave
    </span>
  </div>
</div>
`;
